pants-plugins/macros.py

```diff
@@ -13,6 +13,99 @@
 # limitations under the License.
 
 
+def st2_license(**kwargs):
+    """copy the LICENSE file into each wheel.
+
+    As long as the file is in the src root when building the sdist/wheel,
+    setuptools automatically includes the LICENSE file in the dist-info.
+    """
+    if "dest" not in kwargs:
+        raise ValueError("'dest' path is required for st2_license macro")
+    relocated_files(  # noqa: F821
+        name="license",
+        files_targets=["//:license"],
+        src="",
+        **kwargs,
+    )
+
+
+def st2_runner_python_distribution(**kwargs):
+    runner_name = kwargs.pop("runner_name")
+    description = kwargs.pop("description")
+
+    st2_license(dest=f"contrib/runners/{runner_name}_runner")
+
+    kwargs["provides"] = python_artifact(  # noqa: F821
+        name=f"stackstorm-runner-{runner_name.replace('_', '-')}",
+        description=description,
+        version_file=f"{runner_name}_runner/__init__.py",  # custom for our release plugin
+        # test_suite="tests",
+        zip_safe=kwargs.pop(
+            "zip_safe", True
+        ),  # most runners are safe to run from a zipapp
+    )
+
+    dependencies = kwargs.pop("dependencies", [])
+    for dep in [f"./{runner_name}_runner", ":license"]:
+        if dep not in dependencies:
+            dependencies.append(dep)
+    kwargs["dependencies"] = dependencies
+
+    repositories = kwargs.pop("repositories", [])
+    for repo in ["@pypi"]:
```

Comment:
Why the hardcoded "@pypi" and the loop?

Reply:
Oh. Yes. I never revised that.

The idea here is that we might want to publish wheels to `@testpypi` in addition to `@pypi`, but I never actually included that.

The [`repositories`](https://www.pantsbuild.org/docs/reference-python_distribution#coderepositoriescode) kwarg is:

> List of URL addresses or Twine repository aliases where to publish the Python package.
>
> Twine is used for publishing Python packages, so the address to any kind of repository that Twine supports may be used here.
>
> Aliases are prefixed with `@` to refer to a config section in your Twine configuration, such as a `.pypirc` file. Use `@pypi` to upload to the public PyPi repository, which is the default when using Twine directly.

So, aliases come from [`.pypirc`](https://packaging.python.org/en/latest/specifications/pypirc/#common-configurations) or from the `TWINE_` environment vars. By default, twine defines `@pypi` and `@testpypi`.

Should we simplify this to just check for `@pypi`? Or add `@testpypi` like I originally planned but forgot about?

Reply:
OK. I did some more research. I'll push a commit cleaning this up shortly including a TODO for when we switch to pants 2.16. There's a new `env()` object that will allow switch the repo we publish to using an env var.

Thanks for asking so that I remembered to look into this more.

Reply:
Cleaned up.